### ISSUE TITLE
Figure out non-determinism due to PYTHONHASHSEED

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#243]
Reported by @schmmd.  I'm not really sure what could be causing this, because I didn't think there was any randomness in `model.forward()` after `model.eval()` has been called.  But here are steps to reproduce:

```
$ git checkout schmmd/weird-bug
$ set -x PYTHONHASHSEED 2157
$ allennlp/run serve

> “spaceship”

1.  Navigate to http://localhost:8000.
2.  Click the MC Model tab.
3.  Submit the last example (The Millennium Falcon…)

$ git checkout schmmd/weird-bug
$ set -x PYTHONHASHSEED 4563123
$ allennlp/run serve

1.  Navigate to http://localhost:8000.
2.  Click the MC Model tab.
3.  Submit the last example (The Millennium Falcon…)

> “variety of Star Wars expanded …”
```

**Comment by schmmd** [comment]
Rather than using the web-demo, it's much easier to reproduce using `predict`.

```
allennlp/run predict 'https://s3-us-west-2.amazonaws.com/allennlp/models/bidaf-model-2017.08.26.tar.gz' input.json
```

My branch `schmmd/weird-bug` is [SHA 6bd6e56f1edc855cb1d8428b35705910fb6aff8b](https://github.com/allenai/allennlp/commit/6bd6e56f1edc855cb1d8428b35705910fb6aff8b).

**Comment by schmmd** [comment]
![image](https://user-images.githubusercontent.com/954798/30135837-628f2bbe-9311-11e7-8945-d9d155bd996c.png)

in response to https://github.com/allenai/allennlp/commit/9905e1b680f77d65b0fe2a7a1f8cec641b391002

**Comment by matt-gardner** [comment]
I just tried reproducing this, just to verify that it's been fixed.  I checked out that branch and didn't see this behavior.  I was using python 3.6, though, which fixes dictionary ordering, so maybe that's why?  I then tried creating a python 3.5 environment, and tried again, and I _still_ couldn't get the predictions to switch just by changing `PYTHONHASHSEED`.  Not totally sure what's going on, but I'm going to call this fixed, because the dictionary ordering was almost certainly the cause of the issue.
